design-proposals/klip-28-create-or-replace.md

```diff
@@ -0,0 +1,200 @@
+| Environment | Backfill | Backfill requires the output data to be accurate not just from a point in time, but from the earliest point of retained history |
+| | Cascading | Cascading environments contain queries that are not terminal, but rather feed into downstream stream processing tasks |
+| | Exactly Once | Exactly Once environments do not allow for data duplication or missed events |
+| | Ordered | Ordered environments require that a single offset delineates pre- and post-migration (no events are interleaved) |
+| | Live | Live environments describe queries that cannot afford downtime, either by means of acting as live storage (e.g. responding to pull queries) or feeding into high availability systems (powering important functionality) |
+
+### What is in scope
+
+- Specify a syntax that can support arbitrary upgrades
+- Design a validator to fail unsupported upgrades
+- Design a mechanism for upgrading queries under limited scope
+
+### What is not in scope
+
+The Design section below enumerates which upgrades are out of scope
+
+## Value/Return
+
+This KLIP will represent a significant step forward in the operability of ksqlDB in production, as
+noted in the background and motivation section.
+
+## Public APIS
+
+The syntax `CREATE OR REPLACE (STREAM | TABLE) source_name WITH (key=value, ...) AS query;` will be
+introduced to allow users to specify an existing stream or table to replace with a new query that
+will resume from the same processing point as any previously existing query.
+
+## Design
+
+If the `source_name` does not yet exist, a `CREATE OR REPLACE` statement functions identically to
+a normal `CREATE` statement. Otherwise, ksqlDB executes the following:
+
+1. Identify the original `queryID` that populates the source (`INSERT INTO` discussed later)
+2. Ensure the upgrade is valid
+3. Terminate `queryID`
+4. Start the new query under the same `queryID`
+
+A few changes need to happen in order to make this work. For 1, we need to maintain a mapping from
+source to queryID(s). If the source has multiple associated ids (in the case of `INSERT INTO`) then
+the upgrade will fail and not terminate any queries.
+
+For step 2, there will be a component to determine whether two topologies are "upgrade
+compatible"; the first iterations, which will be delivered as part of this KLIP, will only allow
+for the most basic upgrades: 
+
+- Any _transparent_ upgrade will be supported
+- Any _data selection_ upgrade will be supported
+- _Schema evolution_ upgrades will be supported on simple and stateful queries, but it will be
+    communicated that the users will not get `backfill` or `ordered` properties for stateful.
+- _Source modifying_ upgrades will not be supported
+- _Topology changes_ will not be supported
+
+Note that _Schema Evolution_ compatibility is defined by the limitations of the serialization
+format that is used with the added restrictions against removing fields and changing types to ensure
```

Comment:
Does this mean the only thing that's supported is adding fields? What about reordering fields? Reordering fields would be breaking in the case of DELIMITED.

Reply:
see my note below - that would be a limitation of the serialization format. Reordering would be OK in JSON for example, but not in delimited (and I don't think in AVRO or PROTOBUF either). But essentially the only "useful" query upgrade for schema evolution is adding fields - which I suspect is a vast majority of the use cases for schema evolution.

Reply:
We've even got to be careful when reordering JSON columns.  Once we support multiple key columns, changing the order of key columns would break the binary compatibility of keys.

IMHO, best just to no allow column re-ordering. I can't see it being a much asked for thing anyway.

Reply:
Fair, the ethos of this project is to be extra restrictive and lift restrictions later so we may as well be extra cautious up front.